sympy/functions/special/bessel.py

```diff
@@ -698,64 +698,92 @@ def _eval_as_leading_term(self, x, logx, cdir):
+            return term.as_leading_term(x, logx=logx)
+        elif e.is_negative:
+            # Equation 9.7.2 of Abramowitz and Stegun (10th ed, 1972).
+            return sqrt(pi)*exp(-arg)/sqrt(2*arg)
+        else:
+            return self.func(nu, arg)
 
     def _eval_nseries(self, x, n, logx, cdir=0):
-        # Refer https://functions.wolfram.com/Bessel-TypeFunctions/BesselK/06/01/04/01/02/0008/
-        # for more information on nseries expansion of besselk function.
         from sympy.series.order import Order
         nu, z = self.args
 
-        # In case of powers less than 1, number of terms need to be computed
-        # separately to avoid repeated callings of _eval_nseries with wrong n
         try:
             _, exp = z.leadterm(x)
         except (ValueError, NotImplementedError):
             return self
 
-        if exp.is_positive and nu.is_integer:
-            newn = ceiling(n/exp)
-            bn = besseli(nu, z)
-            a = ((-1)**(nu - 1)*log(z/2)*bn)._eval_nseries(x, n, logx, cdir)
-
-            b, c = [], []
-            o = Order(x**n, x)
+        # In case of powers less than 1, number of terms need to be computed
+        # separately to avoid repeated callings of _eval_nseries with wrong n
+        if exp.is_positive:
             r = (z/2)._eval_nseries(x, n, logx, cdir).removeO()
             if r is S.Zero:
-                return o
-            t = (_mexpand(r**2) + o).removeO()
+                return Order(z**(-nu) + z**nu, x)
 
-            if nu > S.Zero:
-                term = r**(-nu)*factorial(nu - 1)/2
-                b.append(term)
-                for k in range(1, nu):
-                    denom = (k - nu)*k
-                    if denom == S.Zero:
-                        term *= t/k
-                    else:
-                        term *= t/denom
-                    term = (_mexpand(term) + o).removeO()
-                    b.append(term)
+            o = Order(x**n, x)
+            if nu.is_integer:
+                # Reference: https://functions.wolfram.com/Bessel-TypeFunctions/BesselK/06/01/04/01/02/0008/ (only for integer order)
+                newn = ceiling(n/exp)
+                bn = besseli(nu, z)
+                a = ((-1)**(nu - 1)*log(z/2)*bn)._eval_nseries(x, n, logx, cdir)
 
-            p = r**nu*(-1)**nu/(2*factorial(nu))
-            term = p*(digamma(nu + 1) - S.EulerGamma)
-            c.append(term)
-            for k in range(1, (newn + 1)//2):
-                p *= t/(k*(k + nu))
-                p = (_mexpand(p) + o).removeO()
-                term = p*(digamma(k + nu + 1) + digamma(k + 1))
+                b, c = [], []
+                t = (_mexpand(r**2) + o).removeO()
+
+                if nu > S.Zero:
+                    term = r**(-nu)*factorial(nu - 1)/2
+                    b.append(term)
+                    for k in range(1, nu):
+                        denom = (k - nu)*k
+                        if denom == S.Zero:
```

Comment:
@anutosh491 This condition (`denom == S.Zero`) was inherited from your implementation for non-integer order, but it seems to me that it should never be true (by adding an assert, one can at least see that it is not triggered by any of the tests). Can you think of any cases in which it is possible to have `denom == S.Zero`?

Reply:
Hmmm, well if there's a case ... We should surely have a test for it. Will have to look if this would be necessary.

Arnab, can you think of something here. I'll think about it in a while too !

Reply:
I've removed that branch for now (and so we assume `denom != 0`); will add it back if we find an example.